### PR TITLE
Clarify `pl-symbolic-input`'s `allow-sets` interface

### LINKED ISSUE #15628: Clarify the meaning of `pl-symbolic-input[allow-sets]`
### Details

Currently, pl-symbolic-input has a flag for allowing set-type answers (i.e. subclasses of `sympy.Set`, typically `sympy.Interval` and `sympy.FiniteSet`) in addition to the generic `sympy.Expr`. Notably, the `allow-sets` flag _allows_ for sets, but does not make the semantics of _requiring_ them clear. Requiring sets or specific subclasses of sets turns out to be a very common parsing need.

This issue proposes refactoring the attribute into an attribute(s) that make these semantics clear.
- [ ] Required: make requiring `sympy.Set` type expressions explicit
- [ ] Optional: make requiring the common supported subtypes `sympy.Set` possible

### PR DESCRIPTION
**Pull request by SybelBlue**
# Description

**Closes #15628** 

As discussed in the issue, the interface for allowing set-type submissions did not make clear if set-types were required. This PR fixes that by adding recursive type checking as submission post-processing. This allows for recursive structures like Union and Intersection to be checked before returning to the user.

The new API follows Nathan's naming suggestion:
- `allow-sets: bool = 'true'` is unchanged
- `allowed-types: set['all' | 'set' | 'interval' | 'expression'] = {'expression'}` is a new attribute for pl-symbolic-input that restricts output types

This PR was largely drafted with the help of AI, with a majority of implementation written by Codex and then reviewed by @SybelBlue.

- [x] I have disclosed the level of AI assistance used (i.e. none, specific portions, majority of implementation).

# Testing

There are several new tests to cover this behavior:
- end-to-end logic in pl-symbolic-input_test.py
- general function an edge cases in sympy_utils_test.py

### CODE REVIEW

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py [hunk 1]
```diff
@@ -530,6 +531,16 @@ def parse(element_html: str, data: pl.QuestionData) -> None:
...
+        allowed_types=cast(
+            set[psu.AllowedSympyType],
+            set(
+                psu.get_items_list(
+                    pl.get_string_attrib(
+                        element, "allowed-values", ALLOWED_VALUES_DEFAULT
+                    )
+                )
+            ),
+        ),
```
Comment (SybelBlue) [inline apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: The other list type attrs (e.g. additional-simplification) don't go through any validation. I'm happy to add validation before passing on to try_parse...

File: apps/prairielearn/python/test/sympy_utils_test.py [hunk 2]
```diff
@@ -352,6 +352,84 @@ def test_try_parse_string_as_sympy_returns_failure(self) -> None:
...
+        expected: sympy.Expr,
+    ) -> None:
+        result = psu.try_parse_string_as_sympy(
+            text,
+            self.SYMBOL_NAMES,
+            allow_sets=True,
+            allowed_types=allowed_types,
+        )
+
+        assert result == psu.SympyParseSuccess(expected)
```
Comment (SybelBlue) [inline apps/prairielearn/python/test/sympy_utils_test.py:450]: Edge cases for both tests assume that simplification is performed on the output, e.g. `[0, 4] & {1, 2}` is treated as `{1, 2}`. This choice is non-trivial. There is an argument that such intersections should not be allowable under `{"set"}` only conditions. However, this is consistent with my mental model of how our system parses sympy submissions generally. Open to feedback here.

**Review by SybelBlue — COMMENTED** [review]
A couple notes for the reviewer.

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py
Comment (nwalters512) [inline apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Legacy `allow-sets` is used at prepare time to ensure that the correct answer is pareseable under the same rules that will be used when parsing submitted answers. Should we do the same with this new attribute?

Comment (nwalters512) [inline apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Another place where `allow-sets` is used but `allowed-values` is not; this is a smell to me.

File: docs/elements/pl-symbolic-input.md [hunk 3]
```diff
@@ -35,6 +35,7 @@ def generate(data):
...
 | `allow-sets`                    | boolean                 | false                   | Whether set and interval notation are allowed. [See below for more details](#set-notation).                                                                                                                                                                                                                                                     |
+| `allowed-values`                | string                  | `all`                   | Comma-separated types accepted from a submitted answer: `expression`, `set`, or `interval`. `set` accepts only finite sets, including the empty set; `interval` accepts only intervals and the empty set. Use `all` to accept every type of submission.                                                                                         |
```
Comment (nwalters512) [inline docs/elements/pl-symbolic-input.md]: Is the intent for `allowed-values` to supersede `allow-sets`? If so, I think we should just do this here. Grep for `### Migrating from deprecated attributes` to see how other elements handle this.

This actually remained unclear after our discussion. One useful question: is there ever a time when one would want to use `allowed-types="set"` _without_ also setting `allow-sets="true"`? Is that even a valid combination, or should we reject it at runtime during `prepare(...)`?

Comment (nwalters512) [inline docs/elements/pl-symbolic-input.md]: I'd like to bikeshed on the name here. I'm wondering if `allowed-answer-types` would be preferable? "values" is vague in isolation.

File: apps/prairielearn/python/test/sympy_utils_test.py [hunk 4]
```diff
@@ -352,6 +352,84 @@ def test_try_parse_string_as_sympy_returns_failure(self) -> None:
...
+            ("{}", {"interval"}, sympy.EmptySet),
+            ("{}", {"all"}, sympy.EmptySet),
+            ("{1, 2}", {"interval", "set"}, sympy.FiniteSet(1, 2)),
+            ("{1, 2} U {3, 4}", {"set"}, sympy.FiniteSet(1, 2, 3, 4)),
+            ("[1, 2]", {"expression", "interval"}, sympy.Interval(1, 2)),
+            (
+                "[1, 2] U [3, 4]",
+                {"interval"},
+                sympy.Union(sympy.Interval(1, 2), sympy.Interval(3, 4)),
+            ),
```
Comment (nwalters512) [inline apps/prairielearn/python/test/sympy_utils_test.py:391]: This use case (union of intervals) is legit for calculus. Can you anticipate anyone wanting to _only_ accept a single interval, e.g. allow `[0, 2]` but forbid `[1, 2] U [3, 4]`? We can probably add that in a backwards-compatible way later if needed.

File: docs/elements/pl-symbolic-input.md (hunk 3 above)
Comment (nwalters512) [inline docs/elements/pl-symbolic-input.md]: I'd be inclined to do a "see below" for this attribute with more details. The fact that `interval` accepts a union of intervals was non-obvious. Including examples of what qualifies for each would be helpful; those could be lifted straight from the tests.

File: apps/prairielearn/python/test/sympy_utils_test.py
Comment (nwalters512) [inline apps/prairielearn/python/test/sympy_utils_test.py:1]: Codex flagged that `{m} - {n}` is marked as a non-finite set, despite SymPy actually classifying this as a finite set. Likely not relevant for calculus, but could be relevant for discrete math use cases.

File: apps/prairielearn/python/test/sympy_utils_test.py (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/python/test/sympy_utils_test.py:450]: I think this is defensible, yes. However, I think this should also be documented for users.

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py
Comment (nwalters512) [inline apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:105]: As noted on Slack, I believe the new attribute will be rejected at runtime because the schema doesn't include it. I'd recommend adding an example usage of this attribute, perhaps to the existing kitchen-sink question for this element in the example course, and then manually testing that question.

File: apps/prairielearn/python/test/sympy_utils_test.py (hunk 4 above)
Comment (SybelBlue) [reply apps/prairielearn/python/test/sympy_utils_test.py:391]: It's possible, though I haven't had that want. Right now the way to do this would be a typecheck in parse that the answer is a bare `Interval`. Doesn't seem terrible but isn't great either...

Comment (SybelBlue) [reply apps/prairielearn/python/test/sympy_utils_test.py:391]: General Sol. Idea: perform AST comparison before `sympy` simplification with a new passed attribute.
@nwalters512 this may be useful later...

File: docs/elements/pl-symbolic-input.md (hunk 3 above)
Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Migrate from `allow-sets`.

Standup Naming Ideas:
- **`allowed-types`**, don't worry about strict PL/DI sense of "type"
- `answer-types`
- ~~`submission-types`~~, both the correct and submitted answers should be subject to the same validation

Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: For backwards compatibility: `[allow-sets="true"]:not([allowed-types])` -> `allowed-types="all"`
Don't allow both to be set, regardless of compatibility. Error message should point to `allowed-types`.

Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Note to self: make sure the correct answer is also validated, but only by types. Open a new issue for the variables validation problem.

Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Also, rename `set` to `finite-set`.

Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Addressed in [e38d22761](https://github.com/PrairieLearn/PrairieLearn/commit/e38d22761cded0e2585af09a1e353fc7c7650e34).

Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Addressed in [986b31ee9](https://github.com/PrairieLearn/PrairieLearn/commit/986b31ee9eae51d9b94e3d7083a56ca43eaeaccc).

Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Addressed in 46dec959f.

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py
Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Addressed in 0788d9e90. This may cause legacy questions with a disallowed correct-answer type to no longer render; previously, the mismatch may only have failed when parsing a student submission.

Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Follow-up as of ed020adf6: the failure now originates from the new explicit allowed-type validation raises; `convert_string_to_sympy()` once again suffices for parsing in `prepare()`.

Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Addressed in ed020adf6.

File: docs/elements/pl-symbolic-input.md (hunk 3 above)
Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Addressed in d826ea059.

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py
Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:105]: Addressed in e38d22761 and 46dec959f.

File: apps/prairielearn/python/test/sympy_utils_test.py
Comment (SybelBlue) [reply apps/prairielearn/python/test/sympy_utils_test.py:1]: Addressed in d13dc1093.

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py (hunk 1 above)
Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Addressed in e38d22761.

File: apps/prairielearn/python/test/sympy_utils_test.py (hunk 2 above)
Comment (SybelBlue) [reply apps/prairielearn/python/test/sympy_utils_test.py:450]: Addressed in d826ea059.

File: docs/elements/pl-symbolic-input.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SybelBlue) [reply docs/elements/pl-symbolic-input.md]: Addressed in d826ea059. The `allow-sets` dependency was superseded by the migration in 46dec959f; `allowed-types` now enables the required notation directly.

File: apps/prairielearn/python/test/sympy_utils_test.py (hunk 4 above)
Comment (SybelBlue) [reply apps/prairielearn/python/test/sympy_utils_test.py:391]: Retaining the current behavior; unions of intervals remain accepted by `interval` and are documented in d826ea059. A stricter pre-simplification/AST option can be added backwards-compatibly later if needed.

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py [hunk 5]
```diff
@@ -999,12 +1044,17 @@ def test(element_html: str, data: pl.ElementTestData) -> None:
...
-            data["raw_submitted_answers"][name] = f"{random.randint(1, 100):d}"
+        offset = random.randint(1, 100)
+        if not allow_sets and a_tru_str != "":
+            data["raw_submitted_answers"][name] = f"{a_tru_str} + {offset:d}"
+        elif "all" in allowed_types or "expression" in allowed_types:
+            data["raw_submitted_answers"][name] = f"{offset:d}"
+        elif "finite-set" in allowed_types:
+            data["raw_submitted_answers"][name] = f"{{{offset:d}}}"
+        elif "interval" in allowed_types:
+            data["raw_submitted_answers"][name] = f"({offset:d}, {offset + 1:d})"
```
[the remark this answers is not in the document]
Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py]: Addressed in 45b7a46.

File: apps/prairielearn/python/prairielearn/sympy_utils.py [hunk 6]
```diff
@@ -75,6 +75,62 @@ class SympyParseFailure:
...
+    used_types = _used_sympy_types(expr)
+    matches_allowed_type = "all" in allowed_types or (
+        used_types <= allowed_types
+        if used_types
+        else bool({"finite-set", "interval"} & allowed_types)
+    )
+    if matches_allowed_type:
+        return None
+
+    missing_types = (used_types or set()) - allowed_types
```
Comment (austinbillings-pl) [inline apps/prairielearn/python/prairielearn/sympy_utils.py]: It seems in the scenario where the student submits `{ [1, 2] }` where allowed-types="finite-set", this would yield "Your answer has type 'interval', but this input only accepts: finite-set" and submitting `[1, 2] U [3, 4]` would get back "Your answer has type 'non-finite set', but this input only accepts: finite-set"; instead of reporting the missing member type, consider something like:

```suggestion
    missing_types = (used_types or set()) - allowed_types
    if missing_types:
        found = ", ".join(sorted(missing_types))
        return SympyParseFailure(
            f"Your answer uses {found}, which this input does not accept. "
            f"Allowed types: {', '.join(sorted(allowed_types))}."
        )
```
...

File: apps/prairielearn/python/prairielearn/sympy_utils.py [hunk 7]
```diff
@@ -75,6 +75,62 @@ class SympyParseFailure:
...
+                    required_types.update(_used_sympy_types(arg))
+        return required_types
+    if isinstance(expr, (sympy.Union, sympy.Intersection)):
+        required_types: set[AllowedSympyType | Literal["set-base"]] = set()
+        for arg in expr.args:
+            arg_types = _used_sympy_types(arg)
+            required_types.update(arg_types)
+        return required_types
+    if isinstance(expr, sympy.Set):
+        return {"set-base"}
```
Comment (austinbillings-pl) [inline apps/prairielearn/python/prairielearn/sympy_utils.py]: `set-base` isn't part of `AllowedSympyType` so it'll never appear in allowed_types; That makes the `used_types <= allowed_types` eval at line 110 always false for anything in this bucket; as a result, any sympy.Set that isn't an Interval, isn't finite, and isn't a union/Intersection is unconditionally rejected unless the author writes all (e.g. `[0, 5] - {a}`)

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py [hunk 8]
```diff
@@ -136,6 +160,13 @@ def prepare(element_html: str, data: pl.QuestionData) -> None:
                 raise ValueError(
                     f'Parsing correct answer "{a_true}" for "{name}" failed.'
                 ) from exc
+
+            type_failure = psu.check_sympy_types(parsed_answer, allowed_types)
+            if type_failure is not None:
+                raise ValueError(
+                    f'Parsing correct answer "{a_true}" for "{name}" failed: '
+                    f"{type_failure.error}"
+                )
```
Comment (austinbillings-pl) [inline apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:178]: seems that this `has_attrib` branch is the only place the type check is done; if the instructor is setting a correct answer from `server.py` instead then `test()` will fail; I'd suggest moving the check so it runs on the resolved correct answer regardless of source.
My agent had this to say:
> The cleanest shape is to mirror what test() already does: after the correct-answer attribute branch, if name in data["correct_answers"], resolve it (string → convert_string_to_sympy, dict → json_to_sympy) and run check_sympy_types once. That also removes the duplication between prepare() and test().

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 6 above)
Comment (SybelBlue) [reply apps/prairielearn/python/prairielearn/sympy_utils.py]: Addressed in ba0348f69

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 7 above)
Comment (SybelBlue) [reply apps/prairielearn/python/prairielearn/sympy_utils.py]: My solution to this was to go in the other direction: add a "set" type that *is* meaningful. This includes operations with mixed finite-sets/intervals, as well as allowing sympy's special infinite domains like `Naturals`, `Reals` that do not fit the finite-sets/intervals dichotomy. Let me know what you think...

Comment (SybelBlue) [reply apps/prairielearn/python/prairielearn/sympy_utils.py]: Addressed in 8e8a453ab

File: apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py (hunk 8 above)
Comment (SybelBlue) [reply apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:178]: Addressed in 2c3599d11

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 7 above)
Comment (SybelBlue) [reply apps/prairielearn/python/prairielearn/sympy_utils.py]: Note: afaik, because of how sympy handles parsing these names, we allowed "Naturals" and "Reals" to be passed in when the old `allow-sets=True` without declaring them as constants/variables. I never used this in practice, but allowing it explicitly here seemed the most backwards compatible. If compatibility ends up being the only reason to add these, it's worth checking my memory.

**Comment by SybelBlue** [comment]
Major API shift since review: as of 46d4593, the default value is now `{'expression'}` to align with the previous `allow-sets=false` default.

**Comment by SybelBlue** [comment]
# Codex Code Review

I found three issues that should be addressed before merging.

1. **[P1] Preserve the existing `allow_sets=True` parser contract** (`apps/prairielearn/python/prairielearn/sympy_utils.py:1388`). Before this PR, `try_parse_string_as_sympy("{1, 2}", None, allow_sets=True)` returned a successful finite set. The new `allowed_types=None` path silently changes that established call shape to `{"expression"}`, so the same call now fails; `validate_string_as_sympy()` inherits the regression. The existing set-parser tests were changed to pass `allowed_types={"all"}`, which masks the compatibility break instead of testing the old API. When `allowed_types` is omitted, its effective default should follow `allow_sets` (`{"all"}` when true, `{"expression"}` otherwise), while `pl-symbolic-input` can continue passing its explicit policy.

2. **[P1] Keep the pre-existing meaning of SymPy domain names** (`apps/prairielearn/python/prairielearn/sympy_utils.py:576` and `:1188`). On the base branch, `Reals` with `allow_sets=True` and no declaration parses as `sympy.S.Reals`, while an explicitly declared `variables=["Reals"]` remains an ordinary scalar symbol. This PR rejects the first case and rewrites the second to the artificial value `sympy.Set(Symbol("Reals"))`; consequently `Reals + 1` changes from a valid scalar expression to a parse error. It also makes `sympy_to_json(sympy.S.Reals, allow_sets=True)` fail its own round trip with `HasInvalidSymbolError`. That breaks existing `allow-sets="true"` questions even though the documented migration claims `allowed-types="all"` is backward-compatible. The new type classifier can already recognize the real SymPy domain as a set, so these names should retain their old resolution rather than being rejected/wrapped. As a related symptom, the element's `test()` method turns a wrapped domain into `Set(Reals)` at line 1031, then generates a “correct” submission that its own parser rejects because `Set` is not an allowed function.

3. **[P2] Do not make incorrect-answer generation depend only on additive offsets** (`apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:1051`). For valid correct answers such as `infty` and `-infty`, every `a_tru + offset` is equal to `a_tru`; after two iterations this branch raises `AssertionError("Failed to generate an incorrect answer")`. I reproduced this for both infinities. These are common symbolic answers, so question self-testing should fall back to an independently generated expression when addition cannot change the value, with regression coverage for both signs.

Validation: the focused `sympy_utils_test.py` and `pl-symbolic-input_test.py` suites pass (1,154 tests), but neither suite covers the compatibility calls and infinity self-test cases above.

File: apps/prairielearn/python/prairielearn/sympy_utils.py [hunk 9]
```diff
@@ -75,6 +86,67 @@ class SympyParseFailure:
...
 type SympyParseResult = SympyParseSuccess | SympyParseFailure
+type _SympyValueType = Literal["expression", "set", "finite-set", "interval"]
+type _UsedSympyType = _SympyValueType | Literal["empty-set"]
+type AllowedSympyType = Literal["all"] | _SympyValueType
+
+
+def _used_sympy_types(expr: sympy.Basic) -> set[_UsedSympyType]:
+    if expr is sympy.EmptySet:
+        return {"empty-set"}
+    if expr in _SET_DOMAINS:
```
Comment (austinbillings-pl) [inline apps/prairielearn/python/prairielearn/sympy_utils.py]: might we want to check for `sympy.Interval` before the `_SET_DOMAINS` lookup? `(-infty, infty)` parses to an Interval but compares equal to `S.Reals`, so `allowed-types="interval"` rejects it as a set. The same thing happens for `(-infty, 0) U [0, infty)`,  & using either as the correct answer makes `prepare()` fail; these should still be valid interval answers; could we cover both submission and correct-answer validation?

File: apps/prairielearn/python/prairielearn/sympy_utils.py [hunk 10]
```diff
@@ -75,6 +86,67 @@ class SympyParseFailure:
...
+        return {"set"}
+    if isinstance(expr, sympy.Interval):
+        return {"interval"}
+    if isinstance(expr, sympy.Set) and expr.is_finite_set:
```
Comment (austinbillings-pl) [inline apps/prairielearn/python/prairielearn/sympy_utils.py:108]: the finite-set shortcut skips the recursive check when the value is a `Complement` or `Intersection`; With `variables="x,y"` and `allowed-types="finite-set"`, `{[1,2]}` is rejected, but `{[1,2]} - {[x,y]}` and `{[1,2]} & {[x,y]}` are accepted even though the parsed value still contains intervals. could we keep the finite-result classification while checking those nested members?

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 9 above)
Comment (SybelBlue) [reply apps/prairielearn/python/prairielearn/sympy_utils.py]: Addressed in c4edf02988239b10128d1ec4493ccc1cf97d6096.

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 10 above)
Comment (SybelBlue) [reply apps/prairielearn/python/prairielearn/sympy_utils.py:108]: Addressed in 797c1d3241ffdd2efe2fe77aa50126fc5b6c8571.

**Comment by SybelBlue** [comment]
The latest review-fix commit (`797c1d3241ffdd2efe2fe77aa50126fc5b6c8571`) keeps the finite-result classification while checking nested set types. It also adds regression coverage for submissions and correct answers and clarifies the element docs.

**Comment by SybelBlue** [comment]
The last commit adds type overloads that are conditional on passed arguments. I am happy to remove explicit overloads and simply return the generally inclusive value. Let me know what you think.

**Review by austinbillings-pl — APPROVED** [review]
Looks good! One small request; could you just update the PR description for the latest state of the changeset? e.g.`allow-sets` is now deprecated, its old default was `false`, and `finite-set` is missing from the allowed types. It would also be good to mention that the restriction applies to correct answers during `prepare()` in addition to submissions
